### PR TITLE
p2p: fixes

### PR DESCRIPTION
**Pull request by subtly**
fix for -warning in cryptopp.cpp. 
add mutex to host iteration of peers for connect. 
only try to connect to as many peers as is necessary. update unit test.

### CODE REVIEW

**Comment by chriseth** [comment]
Title seems wrong - a PR that fixes itself?

**Comment by subtly** [comment]
@chriseth The title is simply "p2p: fixes". Github UI adds #1276

**Comment by chriseth** [comment]
Oh, yes, sorry.

File: libp2p/Host.cpp [hunk 1]
```diff
@@ -550,14 +550,22 @@ void Host::run(boost::system::error_code const&)
...
 
-	if (peerCount() < m_idealPeerCount)
+	auto openSlots = m_idealPeerCount - peerCount();
+	if (openSlots > 0)
 	{
-		for (auto p: m_peers)
-			if (p.second->shouldReconnect())
-			{
-				connect(p.second);
+		list<shared_ptr<Peer>> toConnect;
```
Comment (chriseth) [inline libp2p/Host.cpp]: Why list?

Comment (subtly) [inline libp2p/Host.cpp]: I don't know how many results will be inserted ahead of time and expect container to have copy.

**Pull request merged**